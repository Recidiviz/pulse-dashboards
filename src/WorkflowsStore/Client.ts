// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { entries, makeObservable, remove, set, when } from "mobx";
import { format as formatPhone } from "phone-fns";

import {
  trackClientProfileViewed,
  trackOpportunityPreviewed,
  trackProfileOpportunityClicked,
  trackReferralFormPrinted,
  trackReferralFormViewed,
  trackSetOpportunityStatus,
  trackSurfacedInList,
} from "../analytics";
import {
  ClientRecord,
  FullName,
  SpecialConditionCode,
  updateOpportunityCompleted,
  updateOpportunityDenial,
} from "../firestore";
import type { RootStore } from "../RootStore";
import {
  CompliantReportingOpportunity,
  EarlyTerminationOpportunity,
  EarnedDischargeOpportunity,
  LSUOpportunity,
  Opportunity,
  OPPORTUNITY_TYPES,
  OpportunityType,
  PastFTRDOpportunity,
} from "./Opportunity";
import { optionalFieldToDate } from "./utils";
import { OTHER_KEY } from "./WorkflowsStore";

export const UNKNOWN = "Unknown" as const;

const SUPERVISION_LEVEL_MAP: Record<string, string> = {
  // TODO(#2287): remove these levels derived from TN raw data
  "STANDARD: MEDIUM": "Medium",
  "STANDARD: MINIMUM": "Minimum",
};

type ClientDetailsCopy = {
  supervisionStartDate?: {
    tooltip: string;
  };
};

export const CLIENT_DETAILS_COPY: Record<string, ClientDetailsCopy> = {
  US_ND: {
    supervisionStartDate: {
      tooltip:
        "This date is included in the filled form as the date that the Defendant was sentenced; in rare cases, the sentencing date and supervision start date may differ. Double check the date in the form to ensure that the sentencing date and supervision start date match in this case.",
    },
  },
};

type OpportunityConfig = {
  flag: `${OpportunityType}Eligible`;
  OpportunityClass: {
    new (client: Client): Opportunity;
  };
};

const OPPORTUNITY_CREATION_MAPPING: Record<
  OpportunityType,
  OpportunityConfig
> = {
  compliantReporting: {
    flag: "compliantReportingEligible",
    OpportunityClass: CompliantReportingOpportunity,
  },
  earlyTermination: {
    flag: "earlyTerminationEligible",
    OpportunityClass: EarlyTerminationOpportunity,
  },
  earnedDischarge: {
    flag: "earnedDischargeEligible",
    OpportunityClass: EarnedDischargeOpportunity,
  },
  LSU: { flag: "LSUEligible", OpportunityClass: LSUOpportunity },
  pastFTRD: { flag: "pastFTRDEligible", OpportunityClass: PastFTRDOpportunity },
};

type OpportunityMapping = {
  earlyTermination?: EarlyTerminationOpportunity;
  compliantReporting?: CompliantReportingOpportunity;
  earnedDischarge?: EarnedDischargeOpportunity;
  LSU?: LSUOpportunity;
  pastFTRD?: PastFTRDOpportunity;
};

export class Client {
  rootStore: RootStore;

  recordId!: string;

  id!: string;

  record!: ClientRecord;

  pseudonymizedId!: string;

  stateCode!: string;

  fullName!: FullName;

  officerId!: string;

  supervisionType!: string;

  supervisionLevelStart?: Date;

  address!: string;

  private rawPhoneNumber?: string;

  expirationDate?: Date;

  supervisionStartDate?: Date;

  currentBalance?: number;

  lastPaymentAmount?: number;

  lastPaymentDate?: Date;

  probationSpecialConditions?: string[];

  paroleSpecialConditions?: SpecialConditionCode[];

  formIsPrinting = false;

  opportunities: OpportunityMapping = {};

  constructor(record: ClientRecord, rootStore: RootStore) {
    makeObservable(this, {
      opportunities: true,
      record: true,
      currentUserEmail: true,
      formIsPrinting: true,
      opportunitiesAlmostEligible: true,
      opportunitiesEligible: true,
      printReferralForm: true,
      setFormIsPrinting: true,
      supervisionLevel: true,
      updateRecord: true,
    });

    this.rootStore = rootStore;

    this.updateRecord(record);
  }

  updateRecord(record: ClientRecord): void {
    this.recordId = record.recordId;
    this.record = record;
    this.id = record.personExternalId;
    this.pseudonymizedId = record.pseudonymizedId;
    this.stateCode = record.stateCode.toUpperCase();
    this.fullName = record.personName;
    this.officerId = record.officerId;
    this.supervisionType = record.supervisionType;
    this.supervisionLevelStart = optionalFieldToDate(
      record.supervisionLevelStart
    );
    this.address = record.address || UNKNOWN;
    this.rawPhoneNumber = record.phoneNumber;
    this.expirationDate = optionalFieldToDate(record.expirationDate);
    this.currentBalance = record.currentBalance;
    this.lastPaymentDate = optionalFieldToDate(record.lastPaymentDate);
    this.lastPaymentAmount = record.lastPaymentAmount;
    this.probationSpecialConditions = record.specialConditions;
    this.paroleSpecialConditions = record.boardConditions ?? [];
    this.supervisionStartDate = optionalFieldToDate(
      record.supervisionStartDate
    );

    OPPORTUNITY_TYPES.forEach((t) => {
      const { flag, OpportunityClass } = OPPORTUNITY_CREATION_MAPPING[t];
      if (record[flag]) {
        if (!this.opportunities[t]) {
          set(this.opportunities, t, new OpportunityClass(this));
        }
      } else {
        remove(this.opportunities, t);
      }
    });
  }

  get displayName(): string {
    return [this.fullName.givenNames, this.fullName.surname]
      .filter((n) => Boolean(n))
      .join(" ");
  }

  get supervisionLevel(): string {
    const levelFromFilters = this.rootStore.workflowsStore.supervisionLevels.find(
      (opt) => opt.value === this.record.supervisionLevel
    )?.label;
    // some levels are mapped locally for historical reasons (e.g., to patch gaps in ingest or handle raw data)
    const locallyMappedLevel =
      SUPERVISION_LEVEL_MAP[this.record.supervisionLevel ?? ""];
    return (
      levelFromFilters ??
      locallyMappedLevel ??
      // TODO(#2287): remove TN-specific override once data is migrated;
      // this prevents a sudden proliferation of "Unknown" labels in TN in the interim
      (this.stateCode === "US_TN" ? "" : UNKNOWN)
    );
  }

  get formattedPhoneNumber(): string | undefined {
    return this.rawPhoneNumber
      ? formatPhone("(NNN) NNN-NNNN", this.rawPhoneNumber)
      : undefined;
  }

  get phoneNumber(): string {
    return this.formattedPhoneNumber || UNKNOWN;
  }

  get officerDistrict(): string | undefined {
    const officer = this.rootStore.workflowsStore?.availableOfficers.find(
      (o) => o.id === this.officerId
    );
    return officer?.district;
  }

  get opportunitiesEligible(): OpportunityMapping {
    return entries(this.opportunities).reduce((opportunities, [key, opp]) => {
      if (opp?.isHydrated && !opp.almostEligible) {
        return { ...opportunities, [key as OpportunityType]: opp };
      }
      return opportunities;
    }, {} as OpportunityMapping);
  }

  get opportunitiesAlmostEligible(): OpportunityMapping {
    return entries(this.opportunities).reduce((opportunities, [key, opp]) => {
      if (opp?.isHydrated && opp.almostEligible) {
        return { ...opportunities, [key as OpportunityType]: opp };
      }
      return opportunities;
    }, {} as OpportunityMapping);
  }

  get currentUserEmail(): string | null | undefined {
    return this.rootStore.workflowsStore.user?.info.email;
  }

  get currentUserName(): string | null | undefined {
    return this.rootStore.workflowsStore.user?.info.email;
  }

  async setOpportunityDenialReasons(
    reasons: string[],
    opportunityType: OpportunityType
  ): Promise<void> {
    if (this.currentUserEmail) {
      // clear irrelevant "other" text if necessary
      const deletions = reasons.includes(OTHER_KEY)
        ? undefined
        : { otherReason: true };

      await updateOpportunityDenial(
        this.currentUserEmail,
        this.recordId,
        { reasons },
        opportunityType,
        deletions
      );

      await updateOpportunityCompleted(
        this.currentUserEmail,
        this.recordId,
        opportunityType,
        true
      );

      if (reasons.length) {
        trackSetOpportunityStatus({
          clientId: this.pseudonymizedId,
          status: "DENIED",
          opportunityType,
          deniedReasons: reasons,
        });
      } else {
        trackSetOpportunityStatus({
          clientId: this.pseudonymizedId,
          status: "IN_PROGRESS",
          opportunityType,
        });
      }
    }
  }

  async setOpportunityOtherReason(
    opportunityType: OpportunityType,
    otherReason?: string
  ): Promise<void> {
    if (this.currentUserEmail) {
      await updateOpportunityDenial(
        this.currentUserEmail,
        this.recordId,
        {
          otherReason,
        },
        opportunityType
      );
    }
  }

  setFormIsPrinting(value: boolean): void {
    this.formIsPrinting = value;
  }

  printReferralForm(opportunityType: OpportunityType): void {
    const opportunity = this.opportunities[opportunityType];
    if (this.currentUserEmail && opportunity?.isHydrated) {
      if (opportunity.reviewStatus !== "DENIED") {
        updateOpportunityCompleted(
          this.currentUserEmail,
          this.recordId,
          opportunityType
        );
        if (opportunity.reviewStatus !== "COMPLETED") {
          trackSetOpportunityStatus({
            clientId: this.pseudonymizedId,
            status: "COMPLETED",
            opportunityType,
          });
        }
      }

      this.setFormIsPrinting(true);
      trackReferralFormPrinted({
        clientId: this.pseudonymizedId,
        opportunityType,
      });
    }
  }

  async trackFormViewed(formType: OpportunityType): Promise<void> {
    await when(() => this.opportunities[formType]?.isLoading === false);

    trackReferralFormViewed({
      clientId: this.pseudonymizedId,
      opportunityType: formType,
    });
  }

  async trackListViewed(listType: OpportunityType): Promise<void> {
    await when(() => this.opportunities[listType]?.isLoading === false);

    trackSurfacedInList({
      clientId: this.pseudonymizedId,
      opportunityType: listType,
    });
  }

  trackProfileViewed(): void {
    trackClientProfileViewed({ clientId: this.pseudonymizedId });
  }

  trackProfileOpportunityClicked(opportunityType: OpportunityType): void {
    trackProfileOpportunityClicked({
      clientId: this.pseudonymizedId,
      opportunityType,
    });
  }

  async trackOpportunityPreviewed(
    opportunityType: OpportunityType
  ): Promise<void> {
    await when(() => this.opportunities[opportunityType]?.isLoading === false);

    trackOpportunityPreviewed({
      clientId: this.pseudonymizedId,
      opportunityType,
    });
  }

  get detailsCopy(): ClientDetailsCopy {
    return CLIENT_DETAILS_COPY[this.stateCode];
  }
}
