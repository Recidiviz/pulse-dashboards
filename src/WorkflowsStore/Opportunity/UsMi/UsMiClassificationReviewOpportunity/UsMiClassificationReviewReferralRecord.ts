/*
 * Recidiviz - a data platform for criminal justice reform
 * Copyright (C) 2023 Recidiviz, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 * =============================================================================
 */

import { z } from "zod";

import {
  caseNotesSchema,
  eligibleDateSchema,
  opportunitySchemaBase,
} from "../../schemaHelpers";

export const usMiClassificationReviewSchemaForSupervisionLevelFormatter = (
  formatter: (raw: string) => string = (s) => s
) =>
  opportunitySchemaBase
    .extend({
      eligibleCriteria: z
        .object({
          usMiNotAlreadyOnLowestEligibleSupervisionLevel: z
            .object({
              supervisionLevel: z.string().transform(formatter).nullable(),
              requiresSoRegistration: z.boolean().nullable(),
            })
            .nullable(),
        })
        .and(
          // this tells zod (and typescript) that we expect exactly one of these fields to be set
          z.union([
            z.object({
              usMiPastInitialClassificationReviewDate: eligibleDateSchema,
              usMiSixMonthsPastLastClassificationReviewDate: z.undefined(),
            }),
            z.object({
              usMiPastInitialClassificationReviewDate: z.undefined(),
              usMiSixMonthsPastLastClassificationReviewDate: eligibleDateSchema,
            }),
          ])
        )
        .transform(
          ({
            usMiPastInitialClassificationReviewDate,
            usMiSixMonthsPastLastClassificationReviewDate,
            ...rest
          }) => ({
            usMiClassificationReviewPastDueDate:
              usMiPastInitialClassificationReviewDate ??
              usMiSixMonthsPastLastClassificationReviewDate,
            ...rest,
          })
        ),
    })
    .merge(caseNotesSchema);

export type UsMiClassificationReviewReferralRecordRaw = z.input<
  ReturnType<typeof usMiClassificationReviewSchemaForSupervisionLevelFormatter>
>;
export type UsMiClassificationReviewReferralRecord = z.infer<
  ReturnType<typeof usMiClassificationReviewSchemaForSupervisionLevelFormatter>
>;
