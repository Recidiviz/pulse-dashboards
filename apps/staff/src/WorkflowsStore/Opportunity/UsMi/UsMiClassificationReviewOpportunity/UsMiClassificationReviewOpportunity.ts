/*
 * Recidiviz - a data platform for criminal justice reform
 * Copyright (C) 2023 Recidiviz, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 * =============================================================================
 */

import { makeObservable } from "mobx";

import { formatWorkflowsDate } from "../../../../utils";
import { Client } from "../../../Client";
import { OpportunityBase } from "../../OpportunityBase";
import { OpportunityRequirement } from "../../types";
import {
  UsMiClassificationReviewReferralRecord,
  usMiClassificationReviewSchemaForSupervisionLevelFormatter,
} from "./UsMiClassificationReviewReferralRecord";

export class UsMiClassificationReviewOpportunity extends OpportunityBase<
  Client,
  UsMiClassificationReviewReferralRecord
> {
  readonly hideUnknownCaseNoteDates = true;

  constructor(client: Client) {
    super(
      client,
      "usMiClassificationReview",
      client.rootStore,
      usMiClassificationReviewSchemaForSupervisionLevelFormatter((raw) =>
        client.rootStore.workflowsStore.formatSupervisionLevel(raw),
      ).parse,
    );

    makeObservable(this, { requirementsMet: true });
  }

  get requirementsMet(): OpportunityRequirement[] {
    if (!this.record) return [];

    const { eligibleDate } =
      this.record.eligibleCriteria.usMiClassificationReviewPastDueDate;

    return [
      {
        text: `Recommended classification review date, based on supervision start date and last classification review date, is ${formatWorkflowsDate(
          eligibleDate,
        )}`,
        tooltip:
          "Classification reviews shall be completed after six months of active supervision […] " +
          "Subsequent classification reviews shall be scheduled at six-month intervals.",
      },

      {
        text: "Currently eligible based on offense type and supervision level",
        tooltip:
          "The supervising Agent shall ensure that a Correctional Offender " +
          "Management Profiling for Alternative Sanctions (COMPAS) has been completed " +
          "for each offender on their active caseload as outlined in OP 06.01.145 " +
          "“Administration and Use of COMPAS and TAP.” Unless mandated by statute or " +
          "other criteria as directed in this operating procedure, the COMPAS shall be " +
          "used to determine the initial supervision level of each offender.\n" +
          "Unless an offender’s supervision level is mandated by policy or statute, " +
          "the supervising Agent shall reduce an offender’s supervision level if " +
          "the offender has satisfactorily completed six continuous months at the " +
          "current assigned supervision level.",
      },
    ];
  }

  get eligibilityDate(): Date | undefined {
    if (!this.record) return;
    return this.record.eligibleCriteria.usMiClassificationReviewPastDueDate
      .eligibleDate;
  }
}
