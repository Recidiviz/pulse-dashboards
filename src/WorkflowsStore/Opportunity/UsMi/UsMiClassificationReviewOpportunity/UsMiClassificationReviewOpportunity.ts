/*
 * Recidiviz - a data platform for criminal justice reform
 * Copyright (C) 2023 Recidiviz, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 * =============================================================================
 */

import { makeObservable } from "mobx";

import { WORKFLOWS_METHODOLOGY_URL } from "../../../../core/utils/constants";
import { OpportunityProfileModuleName } from "../../../../core/WorkflowsClientProfile/OpportunityProfile";
import { formatWorkflowsDate } from "../../../../utils";
import { Client } from "../../../Client";
import { OTHER_KEY } from "../../../utils";
import { OpportunityBase } from "../../OpportunityBase";
import { OpportunityRequirement } from "../../types";
import {
  UsMiClassificationReviewReferralRecord,
  usMiClassificationReviewSchemaForSupervisionLevelFormatter,
} from "./UsMiClassificationReviewReferralRecord";

export class UsMiClassificationReviewOpportunity extends OpportunityBase<
  Client,
  UsMiClassificationReviewReferralRecord
> {
  readonly hideUnknownCaseNoteDates = true;

  readonly opportunityProfileModules: OpportunityProfileModuleName[] = [
    "ClientProfileDetails",
    "CaseNotes",
  ];

  constructor(client: Client) {
    super(
      client,
      "usMiClassificationReview",
      client.rootStore,
      usMiClassificationReviewSchemaForSupervisionLevelFormatter((raw) =>
        client.rootStore.workflowsStore.formatSupervisionLevel(raw)
      ).parse
    );

    makeObservable(this, { requirementsMet: true });
  }

  get requirementsMet(): OpportunityRequirement[] {
    if (!this.record) return [];

    const { eligibleDate } =
      this.record.eligibleCriteria.usMiClassificationReviewPastDueDate;

    return [
      {
        text: `Recommended classification review date, based on supervision start date and last classification review date, is ${formatWorkflowsDate(
          eligibleDate
        )}`,
        tooltip:
          "Classification reviews shall be completed after six months of active supervision […] " +
          "Subsequent classification reviews shall be scheduled at six-month intervals.",
      },

      {
        text: "Currently eligible based on offense type and supervision level",
        tooltip:
          "The supervising Agent shall ensure that a Correctional Offender " +
          "Management Profiling for Alternative Sanctions (COMPAS) has been completed " +
          "for each offender on their active caseload as outlined in OP 06.01.145 " +
          "“Administration and Use of COMPAS and TAP.” Unless mandated by statute or " +
          "other criteria as directed in this operating procedure, the COMPAS shall be " +
          "used to determine the initial supervision level of each offender.\n" +
          "Unless an offender’s supervision level is mandated by policy or statute, " +
          "the supervising Agent shall reduce an offender’s supervision level if " +
          "the offender has satisfactorily completed six continuous months at the " +
          "current assigned supervision level.",
      },
    ];
  }

  readonly policyOrMethodologyUrl = WORKFLOWS_METHODOLOGY_URL.US_MI;

  denialReasonsMap = {
    VIOLATIONS: "Excessive violation behavior during current review period",
    EMPLOYMENT:
      "Chronic unemployment with no effort to job search or recent, concerning unemployment",
    "FINES & FEES":
      "No effort to pay fines and fees despite documented ability to pay",
    "CASE PLAN":
      "No progress toward completion of Transition Accountability Plan goals/tasks",
    NONCOMPLIANT: "Noncompliant with the order of supervision",
    ABSCONSION: "Chronic missing of reporting dates",
    [OTHER_KEY]: "Other: please specify a reason",
  };

  get eligibilityDate(): Date | undefined {
    if (!this.record) return;
    return this.record.eligibleCriteria.usMiClassificationReviewPastDueDate
      .eligibleDate;
  }
}
