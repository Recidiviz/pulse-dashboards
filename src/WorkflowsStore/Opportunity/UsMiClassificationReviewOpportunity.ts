/*
 * Recidiviz - a data platform for criminal justice reform
 * Copyright (C) 2023 Recidiviz, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 * =============================================================================
 */

import { makeObservable } from "mobx";

import { WORKFLOWS_METHODOLOGY_URL } from "../../core/utils/constants";
import { OpportunityProfileModuleName } from "../../core/WorkflowsClientProfile/OpportunityProfile";
import { formatWorkflowsDate } from "../../utils";
import { Client } from "../Client";
import { TransformFunction } from "../subscriptions";
import { fieldToDate, OTHER_KEY } from "../utils";
import { OpportunityBase } from "./OpportunityBase";
import { OpportunityRequirement } from "./types";

export type UsMiClassificationReviewReferralRecord = {
  eligibleCriteria: {
    usMiNotAlreadyOnLowestEligibleSupervisionLevel: {
      supervisionLevel: string;
    };
    usMiClassificationReviewPastDueDate: {
      eligibleDate: Date;
    };
  };
};

export const getRecordTransformer = (client: Client) => {
  const transformer: TransformFunction<
    UsMiClassificationReviewReferralRecord
  > = (record) => {
    if (!record) {
      throw new Error("No record found");
    }

    const { stateCode, externalId, eligibleCriteria } = record;

    // The copy for these eligibility dates are the same. They are different
    // in TES, but we can treat them as the same thing in the frontend
    // since any client will only have one or the other
    let eligibleDate;
    if ("usMiPastInitialClassificationReviewDate" in eligibleCriteria) {
      eligibleDate =
        eligibleCriteria.usMiPastInitialClassificationReviewDate.eligibleDate;
    } else {
      eligibleDate =
        eligibleCriteria.usMiSixMonthsPastLastClassificationReviewDate
          .eligibleDate;
    }

    return {
      stateCode,
      externalId,
      eligibleCriteria: {
        usMiNotAlreadyOnLowestEligibleSupervisionLevel: {
          supervisionLevel:
            client.rootStore.workflowsStore.formatSupervisionLevel(
              eligibleCriteria.usMiNotAlreadyOnLowestEligibleSupervisionLevel
                ?.supervisionLevel
            ),
        },
        usMiClassificationReviewPastDueDate: {
          eligibleDate: fieldToDate(eligibleDate),
        },
      },
    };
  };
  return transformer;
};

export class UsMiClassificationReviewOpportunity extends OpportunityBase<
  Client,
  UsMiClassificationReviewReferralRecord
> {
  readonly opportunityProfileModules: OpportunityProfileModuleName[] = [
    "ClientProfileDetails",
  ];

  constructor(client: Client) {
    super(
      client,
      "usMiClassificationReview",
      client.rootStore,
      getRecordTransformer(client)
    );

    makeObservable(this, { requirementsMet: true });
  }

  get requirementsMet(): OpportunityRequirement[] {
    if (!this.record) return [];

    const { eligibleDate } =
      this.record.eligibleCriteria.usMiClassificationReviewPastDueDate;

    return [
      {
        text: `Recommended classification review date, based on supervision start date and last classification review date, is ${formatWorkflowsDate(
          eligibleDate
        )}`,
        tooltip:
          "Classification reviews shall be completed after six months of active supervision […] " +
          "Subsequent classification reviews shall be scheduled at six-month intervals.",
      },

      {
        text: "Currently eligible based on offense type and supervision level",
        tooltip:
          "The supervising Agent shall ensure that a Correctional Offender " +
          "Management Profiling for Alternative Sanctions (COMPAS) has been completed " +
          "for each offender on their active caseload as outlined in OP 06.01.145 " +
          "“Administration and Use of COMPAS and TAP.” Unless mandated by statute or " +
          "other criteria as directed in this operating procedure, the COMPAS shall be " +
          "used to determine the initial supervision level of each offender.\n" +
          "Unless an offender’s supervision level is mandated by policy or statute, " +
          "the supervising Agent shall reduce an offender’s supervision level if " +
          "the offender has satisfactorily completed six continuous months at the " +
          "current assigned supervision level.",
      },
    ];
  }

  readonly policyOrMethodologyUrl = WORKFLOWS_METHODOLOGY_URL.US_MI;

  denialReasonsMap = {
    VIOLATIONS: "Excessive violation behavior during current review period",
    EMPLOYMENT:
      "Chronic unemployment with no effort to job search or recent, concerning unemployment",
    "FINES & FEES":
      "No effort to pay fines and fees despite documented ability to pay",
    "CASE PLAN":
      "No progress toward completion of Transition Accountability Plan goals/tasks",
    NONCOMPLIANT: "Noncompliant with the order of supervision",
    ABSCONSION: "Chronic missing of reporting dates",
    [OTHER_KEY]: "Other: please specify a reason",
  };

  get eligibilityDate(): Date | undefined {
    if (!this.record) return;
    return this.record.eligibleCriteria.usMiClassificationReviewPastDueDate
      .eligibleDate;
  }
}
