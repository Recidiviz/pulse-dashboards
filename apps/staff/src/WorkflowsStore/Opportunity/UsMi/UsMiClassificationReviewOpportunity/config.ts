// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { add } from "date-fns";
import simplur from "simplur";

import { WORKFLOWS_METHODOLOGY_URL } from "../../../../core/utils/constants";
import { OTHER_KEY } from "../../../utils";
import { OpportunityConfig } from "../../OpportunityConfigs";
import { UsMiClassificationReviewOpportunity } from "./UsMiClassificationReviewOpportunity";

export const usMiClassificationReviewConfig: OpportunityConfig<UsMiClassificationReviewOpportunity> =
  {
    systemType: "SUPERVISION",
    stateCode: "US_MI",
    urlSection: "classificationReview",
    label: "Classification Review",
    snooze: {
      // Ideal behavior is to calculate snooze until based on the date of the last classification review
      // or when they are marked ineligible, whichever is earliest.
      defaultSnoozeUntilFn: (snoozedOn: Date) => add(snoozedOn, { days: 180 }),
    },
    hydratedHeader: (formattedCount) => ({
      eligibilityText: simplur`${formattedCount} client[|s] may be `,
      opportunityText: "eligible for a supervision level downgrade",
      callToAction:
        "Review clients who meet the time threshold for classification review and downgrade supervision levels in COMS.",
    }),
    firestoreCollection: "US_MI-classificationReviewReferrals",
    eligibilityDateText: "Next Classification Due Date",
    sidebarComponents: [
      "UsMiRecommendedSupervisionLevel",
      "EligibilityDate",
      "ClientProfileDetails",
      "CaseNotes",
    ],
    methodologyUrl: WORKFLOWS_METHODOLOGY_URL.US_MI,
    denialReasons: {
      VIOLATIONS: "Excessive violation behavior during current review period",
      EMPLOYMENT:
        "Chronic unemployment with no effort to job search or recent, concerning unemployment",
      "FINES & FEES":
        "No effort to pay fines and fees despite documented ability to pay",
      "CASE PLAN":
        "No progress toward completion of Transition Accountability Plan goals/tasks",
      NONCOMPLIANT: "Noncompliant with the order of supervision",
      ABSCONSION: "Chronic missing of reporting dates",
      [OTHER_KEY]: "Other: please specify a reason",
    },
  };
