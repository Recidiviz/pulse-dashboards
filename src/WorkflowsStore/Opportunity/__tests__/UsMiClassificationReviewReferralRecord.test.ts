// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { identity } from "lodash";

import { Client } from "../../Client";
import { getRecordTransformer } from "../UsMiClassificationReviewOpportunity";

const mockClient = {
  rootStore: {
    workflowsStore: {
      formatSupervisionLevel: identity,
    },
  },
};

test("transform record for initial CR", () => {
  const rawRecord: Record<string, any> = {
    stateCode: "US_MI",
    externalId: "cr-eligible-1",
    eligibleCriteria: {
      usMiNotAlreadyOnLowestEligibleSupervisionLevel: {
        supervisionLevel: "MAXIMUM",
      },
      usMiPastInitialClassificationReviewDate: {
        eligibleDate: "2022-12-12",
      },
    },
  };

  expect(
    getRecordTransformer(mockClient as unknown as Client)(rawRecord)
  ).toMatchSnapshot();
});

test("transform record for six-month CR", () => {
  const rawRecord: Record<string, any> = {
    stateCode: "US_MI",
    externalId: "cr-eligible-2",
    eligibleCriteria: {
      usMiNotAlreadyOnLowestEligibleSupervisionLevel: {
        supervisionLevel: "MAXIMUM",
      },
      usMiSixMonthsPastLastClassificationReviewDate: {
        eligibleDate: "2019-01-12",
      },
    },
  };

  expect(
    getRecordTransformer(mockClient as unknown as Client)(rawRecord)
  ).toMatchSnapshot();
});
