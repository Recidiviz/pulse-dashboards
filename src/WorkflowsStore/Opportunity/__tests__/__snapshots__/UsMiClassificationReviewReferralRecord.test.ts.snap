// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform record for initial CR 1`] = `
Object {
  "caseNotes": Object {
    "Recommended supervision level": Array [
      Object {
        "eventDate": null,
        "noteBody": "MEDIUM",
        "noteTitle": null,
      },
    ],
  },
  "eligibleCriteria": Object {
    "usMiClassificationReviewPastDueDate": Object {
      "eligibleDate": 2022-12-12T00:00:00.000Z,
    },
    "usMiNotAlreadyOnLowestEligibleSupervisionLevel": Object {
      "supervisionLevel": "MAXIMUM",
    },
  },
  "externalId": "cr-eligible-1",
  "stateCode": "US_MI",
}
`;

exports[`transform record for null usMiNotAlreadyOnLowestEligibleSupervisionLevel 1`] = `
Object {
  "caseNotes": undefined,
  "eligibleCriteria": Object {
    "usMiClassificationReviewPastDueDate": Object {
      "eligibleDate": 2023-03-14T00:00:00.000Z,
    },
    "usMiNotAlreadyOnLowestEligibleSupervisionLevel": Object {
      "supervisionLevel": undefined,
    },
  },
  "externalId": "cr-eligible-3",
  "stateCode": "US_MI",
}
`;

exports[`transform record for six-month CR 1`] = `
Object {
  "caseNotes": Object {
    "Recommended supervision level": Array [
      Object {
        "eventDate": null,
        "noteBody": "MEDIUM",
        "noteTitle": null,
      },
    ],
  },
  "eligibleCriteria": Object {
    "usMiClassificationReviewPastDueDate": Object {
      "eligibleDate": 2019-01-12T00:00:00.000Z,
    },
    "usMiNotAlreadyOnLowestEligibleSupervisionLevel": Object {
      "supervisionLevel": "MAXIMUM",
    },
  },
  "externalId": "cr-eligible-2",
  "stateCode": "US_MI",
}
`;
